st2common/st2common/services/config.py

```diff
@@ -0,0 +1,200 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Service functions for managing pack configuration inside the datastore.
+"""
+
+import json
+
+from st2common import log as logging
+from st2common.services import keyvalues as keyvalue_service
+from st2common.constants.keyvalue import USER_SCOPE
+from st2common.constants.keyvalue import SYSTEM_SCOPE
+from st2common.util.crypto import symmetric_decrypt
+from st2common.models.api.keyvalue import KeyValuePairAPI
+from st2common.persistence.keyvalue import KeyValuePair
+from st2common.constants.keyvalue import DATASTORE_KEY_SEPARATOR
+
+__all__ = [
+    'get_datastore_key_prefix_for_pack',
+    'get_datastore_key_name',
+
+    'get_datastore_value_for_config_key',
+    'set_datastore_value_for_config_key',
+
+    'get_datastore_value'
+]
+
+LOG = logging.getLogger(__name__)
+
+# Prefix for datastore items which store config values
+# Full keys follow this format: pack_config.<pack name>.<config key name>
+# For example: pack_config.aws.setup.region
+DATASTORE_CONFIG_KEY_PREFIX = 'pack_config'
+
+
+def get_datastore_key_prefix_for_pack(pack_name):
+    prefix = [DATASTORE_CONFIG_KEY_PREFIX, pack_name]
+    prefix = DATASTORE_KEY_SEPARATOR.join(prefix)
+    return prefix
+
+
+def get_datastore_key_name(pack_name, key_name, user=None):
```

Comment:
I am really confused why you are bringing in the notion of pack scoped config variables again. We explicitly decided not to support it now. 

People can specify only 3 things in the config.yaml file in /opt/stackstorm/configs/${pack}/config.yaml:

k1: v1
k2: {{system.foo}}
k3: {{user.bar}}

With this in mind, we just have the problem of config resolution (lookup). We just need to save the flat list of config item as an object in db after resolution. That can just be a DB model.